oneflow/core/common/data_type.h

```diff
@@ -34,68 +34,110 @@ limitations under the License.
-struct IsFloat16<float16> : std::true_type {};
+}  // namespace std
+
+namespace oneflow {
+
+using float16 = half_float::half;
+
+#define _DEFINE_SPEC(_Trait, _Type, _Value) \
+  template<>                                \
+  struct _Trait<_Type> : std::integral_constant<bool, _Value> {};
 
+// Type Trait: extened IsScalarType
+
+template<typename T>
+struct IsScalarType<
+    T, typename std::enable_if<std::is_same<bfloat16, typename std::remove_cv<T>::type>::value
+                               || std::is_same<float16, typename std::remove_cv<T>::type>::value
 #ifdef WITH_CUDA
+                               || std::is_same<half, typename std::remove_cv<T>::type>::value
+#endif  // WITH_CUDA
+                               >::type>
+    final {
+  static const bool value = true;
+};
 
-template<>
-struct IsFloat16<half> : std::true_type {};
+// Type Trait: IsFloat16
 
+template<typename>
+struct __IsFloat16_helper : std::false_type {};
+_DEFINE_SPEC(__IsFloat16_helper, float16, true)
+#ifdef WITH_CUDA
+_DEFINE_SPEC(__IsFloat16_helper, half, true)
 #endif  // WITH_CUDA
 
 template<typename T>
-struct IsFloat16 : std::false_type {};
+struct IsFloat16
+    : std::integral_constant<bool, (__IsFloat16_helper<typename std::remove_cv<T>::type>::value)> {
+};
 
 // Type Trait: IsFloating
-template<typename T>
-struct IsFloating : std::integral_constant<bool, false> {};
+template<typename>
+struct __IsFloating_helper : std::false_type {};
 
 #define SPECIALIZE_TRUE_FLOATING(type_cpp, type_proto) \
-  template<>                                           \
-  struct IsFloating<type_cpp> : std::integral_constant<bool, true> {};
+  _DEFINE_SPEC(__IsFloating_helper, type_cpp, true)
 OF_PP_FOR_EACH_TUPLE(SPECIALIZE_TRUE_FLOATING, FLOATING_DATA_TYPE_SEQ);
 #undef SPECIALIZE_TRUE_FLOATING
-
-// Type Trait: IsIntegral
+_DEFINE_SPEC(__IsFloating_helper, float16, true)
+#ifdef WITH_CUDA
+_DEFINE_SPEC(__IsFloating_helper, half, true)
+#endif  // WITH_CUDA
 
 template<typename T>
-struct IsIntegral : std::integral_constant<bool, false> {};
+struct IsFloating
+    : std::integral_constant<bool, (__IsFloating_helper<typename std::remove_cv<T>::type>::value)> {
+};
+
+// Type Trait: IsIntegral
+template<typename>
+struct __IsIntegral_helper : std::false_type {};
 
 #define SPECIALIZE_TRUE_INTEGRAL(type_cpp, type_proto) \
-  template<>                                           \
-  struct IsIntegral<type_cpp> : std::integral_constant<bool, true> {};
+  _DEFINE_SPEC(__IsIntegral_helper, type_cpp, true)
 OF_PP_FOR_EACH_TUPLE(SPECIALIZE_TRUE_INTEGRAL, INT_DATA_TYPE_SEQ);
 #undef SPECIALIZE_TRUE_INTEGRAL
 
-// Type Trait: IsUnsignedIntegral
-
 template<typename T>
-struct IsUnsignedIntegral : std::integral_constant<bool, false> {};
+struct IsIntegral
+    : std::integral_constant<bool, (__IsIntegral_helper<typename std::remove_cv<T>::type>::value)> {
+};
+
+// Type Trait: IsUnsignedIntegral
+template<typename>
+struct __IsUnsignedIntegral_helper : std::false_type {};
```

Comment:
```suggestion
namespace details {
struct IsUnsignedIntegralHelper : std::false_type {};
}
```
隐藏在 details 命名空间里，用统一的大小写风格

Reply:
已修改